### PR TITLE
Use openSSL 3.0.x for Python 3.8+

### PR DESCRIPTION
**Pull request by marcelotduarte**

### CODE REVIEW

File: docker/build_scripts/build-openssl.sh [hunk 1]
```diff
@@ -39,11 +39,16 @@ fetch_source ${OPENSSL_ROOT}.tar.gz ${OPENSSL_DOWNLOAD_URL}
...
 make > /dev/null
 make install_sw > /dev/null
 popd
 rm -rf ${OPENSSL_ROOT} ${OPENSSL_ROOT}.tar.gz
 
+# Fix for python configure
+mkdir -p /usr/local/ssl/lib
+if [ -e /usr/local/ssl/lib64 ]; then
+	cp /usr/local/ssl/lib64/lib*.a /usr/local/ssl/lib
+fi
```
Comment (mayeut) [inline docker/build_scripts/build-openssl.sh:52]: This shouldn't be necessary if `--libdir=lib` is passed to config

File: docker/build_scripts/build-openssl.sh [hunk 2]
```diff
@@ -39,11 +39,16 @@ fetch_source ${OPENSSL_ROOT}.tar.gz ${OPENSSL_DOWNLOAD_URL}
 check_sha256sum ${OPENSSL_ROOT}.tar.gz ${OPENSSL_HASH}
 tar -xzf ${OPENSSL_ROOT}.tar.gz
 pushd ${OPENSSL_ROOT}
-./config no-shared --prefix=/usr/local/ssl --openssldir=/usr/local/ssl CPPFLAGS="${MANYLINUX_CPPFLAGS}" CFLAGS="${MANYLINUX_CFLAGS} -fPIC" CXXFLAGS="${MANYLINUX_CXXFLAGS} -fPIC" LDFLAGS="${MANYLINUX_LDFLAGS} -fPIC" > /dev/null
+./config no-afalgeng no-shared --prefix=/usr/local/ssl --openssldir=/usr/local/ssl CPPFLAGS="${MANYLINUX_CPPFLAGS}" CFLAGS="${MANYLINUX_CFLAGS} -fPIC" CXXFLAGS="${MANYLINUX_CXXFLAGS} -fPIC" LDFLAGS="${MANYLINUX_LDFLAGS} -fPIC" > /dev/null
```
Comment (mayeut) [inline docker/build_scripts/build-openssl.sh:42]:
```suggestion
./config no-afalgeng no-shared --prefix=/usr/local/ssl --openssldir=/usr/local/ssl --libdir=lib CPPFLAGS="${MANYLINUX_CPPFLAGS}" CFLAGS="${MANYLINUX_CFLAGS} -fPIC" CXXFLAGS="${MANYLINUX_CXXFLAGS} -fPIC" LDFLAGS="${MANYLINUX_LDFLAGS} -fPIC" > /dev/null
```

Comment (mayeut) [reply docker/build_scripts/build-openssl.sh:42]: Why is config called with `no-afalgeng` ?

**Review by mayeut — CHANGES_REQUESTED** [review]
Thanks for the PR
Mostly waiting for OpenSSL 1.1.1 EOL (2023-09-11) to switch to OpenSSL 3.x
If I'm not mistaken, there were some issues with OpenSSL 3 with older CPython versions. I think CPython 3.7 received some back ports (CPython 3.8 should have full support) but not CPython 3.6 (which is EOL).
I do have a branch locally to test OpenSSL 3.x build.
I went to a dynamic build on that branch so I have a bit more changes than in this PR.

One of the changes missing here is the change to `update_native_dependencies.py` to get the update workflow checking for OpenSSL 3.0.x updates rather than 1.1.1 updates.

File: docker/build_scripts/build-openssl.sh [hunk 3]
```diff
@@ -16,7 +16,7 @@ check_var ${OPENSSL_HASH}
 check_var ${OPENSSL_DOWNLOAD_URL}
 
 OPENSSL_VERSION=${OPENSSL_ROOT#*-}
-OPENSSL_MIN_VERSION=1.1.1
+OPENSSL_MIN_VERSION=3.0.0
```
Comment (mayeut) [inline docker/build_scripts/build-openssl.sh:19]: Why change the minimum OpenSSL version ? Should still be 1.1.1

File: docker/build_scripts/build-openssl.sh (hunk 2 above)
Comment (marcelotduarte) [reply docker/build_scripts/build-openssl.sh]: To avoid a warning.

File: docker/build_scripts/build-openssl.sh (hunk 1 above)
Comment (marcelotduarte) [reply docker/build_scripts/build-openssl.sh]: ok, changed.

**Comment by h-vetinari** [comment]
> I think CPython 3.7 received some back ports (CPython 3.8 should have full support)

Indeed 3.7 is not fully ready for OpenSSL 3.0 AFAIK, but OTOH, it will [become](https://peps.python.org/pep-0537/) end of life this month.

File: docker/Dockerfile [hunk 4]
```diff
@@ -90,8 +90,8 @@ RUN export SQLITE_AUTOCONF_ROOT=sqlite-autoconf-3420000 && \
     manylinux-entrypoint /build_scripts/build-sqlite3.sh
 
 COPY build_scripts/build-openssl.sh /build_scripts/
-RUN export OPENSSL_ROOT=openssl-1.1.1u && \
-    export OPENSSL_HASH=e2f8d84b523eecd06c7be7626830370300fbcc15386bf5142d72758f6963ebc6 && \
+RUN export OPENSSL_ROOT=openssl-3.0.8 && \
+    export OPENSSL_HASH=6c13d2bf38fdf31eac3ce2a347073673f5d63263398f1f69d0df4a41253e4b3e && \
```
Comment (h-vetinari) [inline docker/Dockerfile:94]: There's 3.0.9 now which fixes a couple of not-so-critical CVEs that 3.0.8 was exposed to.

Comment (mayeut) [reply docker/Dockerfile]: bumped to 3.0.10

**Comment by marcelotduarte** [comment]
@mayeut Sorry, I didn't have time to continue this PR. For me personally, it would be interesting to have openssl3 in Python 3.10+, but I see that in CPython, for macOS and Windows they only release installers for 3.11+ (even because they no longer produce installers for version 3.10, which is under security fix).
https://github.com/python/cpython/issues/107565

> > I think CPython 3.7 received some back ports (CPython 3.8 should have full support)
> 
> Indeed 3.7 is not fully ready for OpenSSL 3.0 AFAIK, but OTOH, it will [become](https://peps.python.org/pep-0537/) end of life this month.

In CPython there are tests since 3.8:
https://github.com/python/cpython/pull/108124

Are you going to finalize the necessary changes?

File: docker/build_scripts/build-openssl.sh (hunk 3 above)
Comment (auvipy) [reply docker/build_scripts/build-openssl.sh:19]: should we still keep the minimum version same as old? and keep support through 1.x-3.x series?

Comment (marcelotduarte) [reply docker/build_scripts/build-openssl.sh:19]: I understand that there will have to be two versions of openssl, the legacy and the current one. And in the case of the minimum version, it must be the one requested for each case.

**Comment by marcelotduarte** [comment]
@mayeut I will make the changes by tomorrow at the latest.

File: docker/build_scripts/build-openssl.sh (hunk 3 above)
Comment (mayeut) [reply docker/build_scripts/build-openssl.sh]: No, the minimum version, even for python 3.12 is still 1.1.1
This requirement started with python 3.10 and to simplify the workflow, it was set the same for all pythons versions.
So this shall remain 1.1.1 even for 3.8+

Comment (marcelotduarte) [reply docker/build_scripts/build-openssl.sh]: In the recent version of the PR, I reverted the value.

**Comment by marcelotduarte** [comment]
@mayeut Can I squash and rebase the PR?

**Comment by mayeut** [comment]
@marcelotduarte,

I will open a PR to use system openssl for python 3.6 & 3.7 (it will downgrade openssl from 1.1.1 to 1.0.2 for python 3.6 & 3.7 but given they will still be functional and are EOL, I think that's completely fine).

Please wait a bit to rebase in order to rebase on this PR.
I'll let the PR opened about 1-2 weeks in order to give a chance to downstream users to comment.
The other option would  be to build 3.6 & 3.7 with openssl 3.x as well (in case system does not provide 1.1.1) given it seems to pass some basic functionality (i.e., pip install).

File: docker/Dockerfile [hunk 5]
```diff
@@ -90,9 +90,11 @@ RUN export SQLITE_AUTOCONF_ROOT=sqlite-autoconf-3430100 && \
     manylinux-entrypoint /build_scripts/build-sqlite3.sh
 
 COPY build_scripts/build-openssl.sh /build_scripts/
-RUN export OPENSSL_ROOT=openssl-1.1.1w && \
-    export OPENSSL_HASH=cf3098950cb4d853ad95c0841f1f9c6d3dc102dccfcacd521d93925208b76ac8 && \
+RUN export OPENSSL_ROOT=openssl-3.0.10 && \
+    export OPENSSL_HASH=1761d4f5b13a1028b9b6f3d4b8e17feb0cedc9370f6afe61d7193d2cdce83323 && \
     export OPENSSL_DOWNLOAD_URL=https://www.openssl.org/source && \
+    export OPENSSL_ROOT_LEGACY=openssl-1.1.1w && \
+    export OPENSSL_HASH_LEGACY=cf3098950cb4d853ad95c0841f1f9c6d3dc102dccfcacd521d93925208b76ac8 && \
```
Comment (mayeut) [inline docker/Dockerfile:97]: to be removed, I'd rather not build 2 openssl versions if we can avoid it.

Comment (marcelotduarte) [reply docker/Dockerfile:97]: So, I think you don't need to open another PR, I can remove the existing SSL check, so the system's SSL version remains, and for Python, the downloaded version is used. I used a scheme similar to the Python doc: https://docs.python.org/3.10/using/unix.html#custom-openssl

**Review by mayeut — CHANGES_REQUESTED** [review]
I think all commits related to dual openssl versions shall be removed.
The only thing that had an action in the previous review was to restore the minimum version to 1.1.1

File: docker/Dockerfile (hunk 5 above)
Comment (mayeut) [reply docker/Dockerfile:97]: I'd rather be explicit than not, and once those EOL versions are removed, then just remove those blocks of code in Dockerfile.

File: docker/build_scripts/build-cpython.sh [hunk 6]
```diff
@@ -39,16 +39,22 @@ if [ "${CPYTHON_VERSION}" == "3.6.15" ]; then
...
 fi
 if [ "${AUDITWHEEL_POLICY}" == "manylinux2014" ] ; then
-    # Python 3.11+
+	# Python 3.11+
 	export TCLTK_LIBS="-ltk8.6 -ltcl8.6"
 fi
+if [ "${CPYTHON_VERSION%.*}" == "3.6" ] || [ "${CPYTHON_VERSION%.*}" == "3.7" ]; then
+	OPENSSL_OPTIONS=""
+else
+	OPENSSL_OPTIONS="--with-openssl=/usr/local/openssl3"
```
Comment (mayeut) [inline docker/build_scripts/build-cpython.sh:48]: This won't work if the system openssl is good enough as /usr/local/openssl3 won't exist.
Am I missing something ?

File: tests/run_tests.sh [hunk 7]
```diff
@@ -35,6 +35,12 @@ for PYTHON in /opt/python/*/bin/python; do
 		$PYTHON -c 'import sqlite3; print(sqlite3.sqlite_version); assert sqlite3.sqlite_version_info[0:2] >= (3, 34)'
 		# Make sure tkinter module can be loaded properly
 		$PYTHON -c 'import tkinter; print(tkinter.TkVersion); assert tkinter.TkVersion >= 8.6'
+		# Make sure openssl module can be loaded properly
+		if [ "${PYVERS}" == "3.6" ] || [ "${PYVERS}" == "3.7" ]; then
+			$PYTHON -c 'import ssl; print(ssl.OPENSSL_VERSION); assert ssl.OPENSSL_VERSION_INFO >= (1, 1, 1)'
+		else
+			$PYTHON -c 'import ssl; print(ssl.OPENSSL_VERSION); assert ssl.OPENSSL_VERSION_INFO >= (3, 0)'
+		fi
```
Comment (mayeut) [inline tests/run_tests.sh:43]: Those checks will all fail:
system openssl is 1.0.2 on manylinux2014 for 3.6 & 3.7
system openssl is 1.1.1 on manylinux_2_28 so it shall be used for all python versions on that image.

Comment (mayeut) [reply tests/run_tests.sh:43]: PS: the ssl-check.py a couple lines before makes sure that ssl can be imported and is working.
PS2: if you want to add a check on the version, it should be >= 1.1.1 for python >= 3.8 (I guess/hope that PyPy3.8+ is using something recent)

File: docker/build_scripts/build-cpython.sh (hunk 6 above)
Comment (marcelotduarte) [reply docker/build_scripts/build-cpython.sh:49]: Fixed.

File: tests/run_tests.sh (hunk 7 above)
Comment (marcelotduarte) [reply tests/run_tests.sh]: I changed build-openssl and changed to change the other modules. Fixed.
This check is for version only.

**Comment by pelson** [comment]
> I will open a PR to use system openssl for python 3.6 & 3.7 (it will downgrade openssl from 1.1.1 to 1.0.2 for python 3.6 & 3.7 but given they will still be functional and are EOL, I think that's completely fine).

Just a comment to say that this caught me out last week. I put some details in https://github.com/pypa/manylinux/pull/1535#issuecomment-1739181866.

I think calling Python 3.7 EoL is totally reasonable, but it is difficult to call the Python 3.7 shipped as fully "functional". I also appreciate that the case in the attached comment relates to a third-party package (albeit the second-most popular, by download, package on PyPI).

**Pull request closed without merging**